### PR TITLE
Fix issue 255

### LINKED ISSUE #255: Handle 'mid-growth stage' ["model"]
### Summary
The model was built in the context of the first few days of local outbreak. It is not well equipped to handle projections when there is a non-small number of current infections. There needs to be a way for the user/analyst to specify not only how many cases are currently hospitalized, but also how long it has been since the first case was hospitalized.  In keeping with the goals of the application being as simple as possible while providing actionable projections in terms of hospital-level resource demands, the handling of 'mid-stage' starting points should be kept likewise simple.

### Additional details

After the first week or so of spread, there is additional information contained in the combination of current hospitalized case counts and the number of days since the first case. These two quantities should be combined to provide an estimate of local doubling-time. 

### Suggested fix

- Add an input for 'Days since first hospitalized COVID-19 case' {t_{firstcase}}.
- Start SIR sim from that many days back
- Run argmin_{doublingtime} [abs(hospitalized census - predicted)] at time 0 + t_{fistcase}

### PR DESCRIPTION
**Pull request by quinn-dougherty**
Closes https://github.com/CodeForPhilly/chime/issues/255

[notebook for reference](https://github.com/CodeForPhilly/chime/blob/fithxdoublingtime/notebooks/Fit%20doubling%20time%20using%20time%20since%20first%20hospitalization.ipynb)

found a clean way to do it in streamlit's UI that doesn't build up ugly technical debt. 

TODO: need to rearrange the order in which computations happen in order to get it to actually work. in the notebook, inferred doubling time is a function of assumed doubling time, so i just need to make that work with autopopulating the form with a default (and not write myself into literal circles)

### CODE REVIEW

File: src/penn_chime/presentation.py [hunk 1]
```diff
@@ -268,7 +276,12 @@ def display_sidebar(st, d: Constants) -> Parameters:
 
     st.sidebar.markdown("### Spread and Contact Parameters [ℹ]({docs_url}/what-is-chime/parameters)"
                         .format(docs_url=DOCS_URL))
-    doubling_time = doubling_time_input()
+    if st.sidebar.checkbox("I know the number of days since first hospitalized case"):
```
Comment (PhilMiller) [inline src/penn_chime/presentation.py]: Please make this "I know the admission date of the first hospitalized case", so that #146 is easier, and so users don't have to repeat date arithmetic with every use, or accidentally forget to update the number on their sticky note, whiteboard, config file, or whatever.

**Comment by quinn-dougherty** [comment]
> inferred doubling time is a function of assumed doubling time

I was mistaken about this, sorry, correcting approach.

File: src/penn_chime/presentation.py (hunk 1 above)
Comment (quinn-dougherty) [reply src/penn_chime/presentation.py]: that's a good idea! saw your comment in the issue. i'd like to try to squeeze it in this PR! gonna work on bigger problems i'm having first.

Comment (quinn-dougherty) [reply src/penn_chime/presentation.py]: covered this.

**Comment by quinn-dougherty** [comment]
Re Phil's comment. (streamlit's `st.date_input`) 

Screenshot from 2020-03-26 17-17-03

Screenshot from 2020-03-26 17-17-16

**Comment by cjbayesian** [comment]
@quinn-dougherty That looks perfect. Just fix the typo -> "I know the date of the first [...]" (remove 'number of')

File: src/penn_chime/models.py [hunk 2]
```diff
@@ -44,6 +44,14 @@ def __init__(self, p: Parameters) -> SimSirModel:
             p.known_infected / infected if infected > 1.0e-7 else None
         )
 
+
+        #self.hospitalized_t = p.current_hospitalized
+        #self.n_days_since_first_hospitalized = p.n_days_since_first_hospitalized
+        if p.n_days_since_first_hospitalized is not None:
+            # do not shorten this to `if foo:` form. if you do and n_days_since_first_hospitalized is 0,
+            # program will crash.
+            p.doubling_time = self.argmin_dt(np.linspace(1, 15, 29), p)
```
Comment (PhilMiller) [inline src/penn_chime/models.py]: Would it make sense to fit to a finer precision than seemingly half days? I feel like this could introduce some wild discretization error in the forecasts. If there's a performance concern in increasing the precision for the whole range, you can iteratively refine the fit, by doing successive `argmin_dt` calls in shrinking neighborhoods around the previous value found.

Comment (PhilMiller) [inline src/penn_chime/models.py]: This is a really horrifying thought, but is there a principled reason the doubling time can never be less than 1 day? Super-spreader events would seem to do exactly that.

File: src/penn_chime/models.py [hunk 3]
```diff
@@ -97,6 +106,27 @@ def __init__(self, p: Parameters) -> SimSirModel:
         self.daily_growth_t = daily_growth_helper(doubling_time_t)
 
 
+    def observed_predicted_loss(self, doubling_time: float, p: Parameters) -> float:
```
Comment (PhilMiller) [inline src/penn_chime/models.py]: The `doubling_time` parameter isn't actually used in this method

Comment (PhilMiller) [reply src/penn_chime/models.py]: Was the intention to set it in the model, and run the simulation accordingly, and then look at `self.census_df`?

Comment (quinn-dougherty) [reply src/penn_chime/models.py]: thanks! these are details that will be radically different in the fix.

File: src/penn_chime/models.py (hunk 2 above)
Comment (quinn-dougherty) [reply src/penn_chime/models.py]: that's a hardcoded constant from Corey's notebook that we can factor into the `settings.cfg` file if we wish. since we're pushing to prod weekly, we only hope that things don't get that drastic too quickly ;)

Comment (quinn-dougherty) [reply src/penn_chime/models.py]: agreed, but not in the first iteration

File: src/penn_chime/models.py (hunk 3 above)
Comment (PhilMiller) [reply src/penn_chime/models.py]: Sorry, corrected typo mode->model

File: src/penn_chime/models.py (hunk 2 above)
Comment (PhilMiller) [reply src/penn_chime/models.py]: More broadly, if the parameter sweep from `argmin_dt` returns a value at either extreme of its range, it means either
- The best fit value is approximately the given limit
- The fit failed because the true result is outside the range

We need to not let the latter case go through without warning or examination. It may just as well be an indicator of mis-keyed inputs, as of the math honestly saying anything about the circumstances.

Comment (PhilMiller) [reply src/penn_chime/models.py]: There's a physical limit on doubling time down around 0.1, since that implies 10 daily doublings, or about 1000x, which exhausts itself pretty damn quickly. I think there's some threshold higher than that where we should warn the user that the computed value is implausibly low

**Comment by PhilMiller** [comment]
A UI note for as this progresses: The number of days for which the user asks for a projection needs to be in addition to days in the back-cast - i.e. the admissions and census charts should run from `-1*n_days_since_first_hospitalized` to `n_days`

**Comment by quinn-dougherty** [comment]
@cjbayesian 
your notebook has `S, I, R = population_size-initial_i, initial_i, 0`, your `susceptible` population is `total_population - initial_infected`. 

This is not parity with prod. The codebase considers the entire `total_population` susceptible. 

how should we proceed? `initial_i` (`infected` as we call it in prod) is 3 orders of magnitude smaller than total population, so it shouldn't be effecting our results right now.

**Comment by PhilMiller** [comment]
That would be #211, which I realized I need to go back and revise the fix in #212 for the big refactoring.

File: src/penn_chime/models.py (hunk 2 above)
Comment (quinn-dougherty) [reply src/penn_chime/models.py]: i think you're right. i'll add in some sort of warning system.

**Comment by cjbayesian** [comment]
Yep, @PhilMiller is right that I had fixed the susceptible calculation in the notebook per issue #211 .

**Comment by quinn-dougherty** [comment]
>That would be #211, which I realized I need to go back and revise the fix in #212 for the big refactoring.
>Yep, @PhilMiller is right that I had fixed the susceptible calculation in the notebook per issue #211.

https://github.com/CodeForPhilly/chime/pull/276 merged

**Comment by cjbayesian** [comment]
@PhilMiller and @quinn-dougherty : I have a coherent way to handle either case of user input (doubling_time or n_days_since_first_hospitalized) which addresses also #187 . 

I just pushed an update to my notebook showing how. It's basically the same fitting procedure, just holding `doubling_time` constant and fitting `n_days_since_first_hospitalized` instead.

**Comment by quinn-dougherty** [comment]
Per Phil's comments, someone remind us to have a talk about what this `np.linspace(1,15,29)` should be set to before push to prod monday.

**Comment by quinn-dougherty** [comment]
- not using vectorized argmin because i'm keeping track of state. 
- changed `obs_pred_diff` to `loss_dt` 

pushing some commits without the 187 piggyback shortly

**Comment by quinn-dougherty** [comment]
> I have a coherent way to handle either case of user input (doubling_time given n_days_since_first_hospitalized) which addresses also #187 .
I just pushed an update to my notebook showing how. It's basically the same fitting procedure, just holding doubling_time constant and fitting n_days_since_first_hospitalized instead.

sorry, this is gonna have to be another PR. racing to get the basics of 255 into Prod monday.

File: src/penn_chime/models.py (hunk 3 above)
Comment (quinn-dougherty) [reply src/penn_chime/models.py]: out of date per new commits.

**Comment by quinn-dougherty** [comment]
**there's a math error** -- Corey's notebook infers a doubling time of `4` with similar initial conditions, my implementation infers a doubling time of 15 :(

I'll also recall Phil's warning, it's reporting the 15th day (*the max of the range*) as the doubling time on input `03-09-2020` (with checkbox "i know date of first hospitalized") 
Screenshot from 2020-03-27 11-35-43

what Phil wrote: 
> More broadly, if the parameter sweep from argmin_dt returns a value at either extreme of its range, it means either
- The best fit value is approximately the given limit
- The fit failed because the true result is outside the range
We need to not let the latter case go through without warning or examination. It may just as well be an indicator of mis-keyed inputs, as of the math honestly saying anything about the circumstances.

**Comment by quinn-dougherty** [comment]
TODO: 
- implement warning if inferred doubling time is at the top or the bottom of its search space.

File: src/penn_chime/models.py (hunk 2 above)
Comment (jlubken) [reply src/penn_chime/models.py]: I believe that a different merge allowed doubling time to be a float/fractional.

File: src/penn_chime/models.py [hunk 4]
```diff
@@ -94,10 +97,128 @@ def __init__(self, p: Parameters) -> SimSirModel:
...
+            r_t = beta / gamma * susceptible
+            r_naught = (intrinsic_growth_rate + gamma) / gamma
+            doubling_time_t = 1.0 / np.log2(beta * susceptible - gamma + 1)
+            raw_df = sim_sir_df(
+                susceptible,
+                infected,
+                recovered,
+                beta,
+                gamma,
+                p.n_days
```
Comment (jlubken) [inline src/penn_chime/models.py]: Shouldn't this be p.n_days + p.n_days_since_first_hospitalization?

File: src/penn_chime/models.py [hunk 5]
```diff
@@ -94,10 +97,128 @@ def __init__(self, p: Parameters) -> SimSirModel:
...
+            )
+            dispositions_df = build_dispositions_df(raw_df, rates, p.market_share)
+            admits_df = build_admits_df(dispositions_df, p.n_days_since_first_hospitalized)
```
Comment (jlubken) [inline src/penn_chime/models.py]: ... then maybe the p.n_days_since_first_hospitalization doesn't need to be passed in here for admits or for the census?

File: src/penn_chime/models.py (hunk 4 above)
Comment (quinn-dougherty) [reply src/penn_chime/models.py]: all i'll say is that wouldn't be parity with the notebook. @cjbayesian

**Comment by quinn-dougherty** [comment]
hi -- are we ok if i piggyback updating `settings.py` in here? i'm trying to achieve parity with the notebook. i'm assuming it'll happen before monday anyway.

File: src/penn_chime/models.py [hunk 6]
```diff
@@ -136,6 +257,19 @@ def sim_sir_df(
         columns=("day", "susceptible", "infected", "recovered"),
     )
 
+
+def get_dispositions(
```
Comment (jlubken) [inline src/penn_chime/models.py]: This function should not be needed. It was replaced with build_dispositions_df that includes the days column.

File: src/penn_chime/models.py (hunk 4 above)
Comment (jlubken) [reply src/penn_chime/models.py]: OK, so we'll fix this later then. We really want the raw_df to hold all of the data instead of having to bodge n_days_since... into every function.

I'll get a copy of the notebook and test for equivalent output later.

**Comment by jlubken** [comment]
It looks like the cli will be busted. Some work will be needed to allow null for some in the validator.

File: src/penn_chime/models.py (hunk 4 above)
Comment (jlubken) [reply src/penn_chime/models.py]: However, this input can no longer be described as the number of days to project as stated in the ui. It is projecting less than n_days, it is projecting from the first infection. The ui / users here will be confusing / confused.

**Review by jlubken — COMMENTED** [review]
This impacts clarity of the UI. We should file more issues to address these limitations.

Or update pass n_days + n_days_since_first though the raw_df.

This will be confusing to users if we release as is on Monday.

**Comment by quinn-dougherty** [comment]
> It looks like the cli will be busted. Some work will be needed to allow null for some in the validator.

oof. 

presentation.py isn't a good spot for such a missioncritical assignment, but 
https://github.com/CodeForPhilly/chime/blob/quinn_issue_255/src/penn_chime/presentation.py#L281 lines 281-286 are what ensures that if one of them is None, the other one is a value, and vice versa. 

it's true that https://github.com/CodeForPhilly/chime/issues/166 will be harder now.

**Comment by quinn-dougherty** [comment]
TODO: fix tables/csvs -- "negative days" / starting from start date isn't implemented for them

File: src/penn_chime/models.py (hunk 6 above)
Comment (quinn-dougherty) [reply src/penn_chime/models.py]: correct -- was just triplechecking my parity with notebook, but in the current push function is no longer being called

**Comment by quinn-dougherty** [comment]
> OK, so we'll fix this later then. We really want the raw_df to hold all of the data instead of having to bodge n_days_since... into every function.

you're correct. `raw_df.index_of(0)` or `raw_df.index_of(today)` could convey the same information.

**Comment by quinn-dougherty** [comment]
> A UI note for as this progresses: The number of days for which the user asks for a projection needs to be in addition to days in the back-cast - i.e. the admissions and census charts should run from -1*n_days_since_first_hospitalized to n_days

implemented for when `i know the start date` is set to true. can't implement for the opposite case until i roll another argmin (which we're talking about putting in a separate PR after monday), so it looks like i'm going to de-implement this for consistency (should be about one line, which i'll document)

**Comment by PhilMiller** [comment]
> Per Phil's comments, someone remind us to have a talk about what this `np.linspace(1,15,29)` should be set to before push to prod monday.

As a first cut, it should be reasonable to do `np.linspace(0.1, 15, 299)` to get down to 0.1 precision by brute force. If performance takes too much of a hit, then it's worthwhile to implement the refinement approach.

**Comment by quinn-dougherty** [comment]
update: it appears that 187 is piggybacked afterall :gritty2:

**Comment by quinn-dougherty** [comment]
from slack 
> in my test runs, it infers n_days_since_first_hospitalized at 4. seems a little off

>Agreed. It should be more like 20
And when I put in the date of the first hospitalized case, the census at day zero is 218, instead of the ~32 that it should be
Instead, it hits the right census around -15 (~5 days from the start of time)

numbers are still slightly off

**Comment by quinn-dougherty** [comment]
> However, this input can no longer be described as the number of days to project as stated in the ui. It is projecting less than n_days, it is projecting from the first infection. The ui / users here will be confusing / confused.

you're right this is technically a problem. 

but we've always done this with an erroneous `n_days - 10` that i never understood

**Comment by quinn-dougherty** [comment]
closes https://github.com/CodeForPhilly/chime/issues/168

**Comment by cjbayesian** [comment]
@quinn-dougherty & @jlubken See branch`issue_255_corey`. I fixed the calculations and fitting procedure. Needs cleanup. Also still needs to have social distancing beta from day 0 forward added.

**Review by jlubken — APPROVED** [review]
